packages/react-query/src/shared/proxy/utilsProxy.ts

```diff
@@ -332,3 +336,48 @@ export function createReactQueryUtils<TRouter extends AnyRouter, TSSRContext>(
     });
   });
 }
+
+/**
+ * @internal
+ */
+export function createQueryUtilsProxy<TRouter extends AnyRouter>(
+  context: TRPCQueryUtils<TRouter>,
+) {
+  type CreateQueryUtilsProxyReturnType = CreateQueryUtils<TRouter>;
+
+  return createFlatProxy<CreateQueryUtilsProxyReturnType>((key) => {
+    return createRecursiveProxy((opts) => {
+      const path = [key, ...opts.path];
+      const utilName = path.pop() as keyof AnyDecoratedProcedure;
+      const args = [...opts.args] as Parameters<
+        AnyDecoratedProcedure[typeof utilName]
+      >;
+      const input = args.shift(); // args can now be spread when input removed
+      const queryType = getQueryType(utilName);
+      const queryKey = getQueryKeyInternal(path, input, queryType);
+
+      const contextMap: Record<keyof AnyDecoratedProcedure, () => unknown> = {
+        fetch: () => context.fetchQuery(queryKey, ...args),
+        fetchInfinite: () => context.fetchInfiniteQuery(queryKey, args[0]),
+        prefetch: () => context.prefetchQuery(queryKey, ...args),
+        prefetchInfinite: () =>
+          context.prefetchInfiniteQuery(queryKey, args[0]),
+        ensureData: () => context.ensureQueryData(queryKey, ...args),
+        invalidate: () => context.invalidateQueries(queryKey, ...args),
+        reset: () => context.resetQueries(queryKey, ...args),
+        refetch: () => context.refetchQueries(queryKey, ...args),
+        cancel: () => context.cancelQuery(queryKey, ...args),
+        setData: () => {
+          context.setQueryData(queryKey, args[0], args[1]);
+        },
+        setInfiniteData: () => {
+          context.setInfiniteQueryData(queryKey, args[0], args[1]);
+        },
+        getData: () => context.getQueryData(queryKey),
+        getInfiniteData: () => context.getInfiniteQueryData(queryKey),
+      };
```

Comment:
we already have this somewhere, should be able to reuse it rigth?

Reply:
oh it's right above here

Reply:
> oh it's right above here, this looks scary similar so can most likely get rid of one?
> 
> https://github.com/trpc/trpc/blob/5f66bfba30b32ecab6237cbd4402c63ff5d6ba2c/packages/react-query/src/shared/proxy/utilsProxy.ts#L287-L383

It looks like there's some SSR functionality embedded in the other one which is why I haven't just reused a bunch of stuff from the other utils proxy.

Reply:
@juliusmarminge I think I've made it as reusable as possible in the commit following this conversation.